# Summary

## Data pre-processing and imputations 

1. Web scraping was done on domain.com.au to obtain listing level data. A public endpoint was utilised that returned a json payload with all the required information about the current listed porperty such as price, name, address, bed, bath...etc. Further external data was also obtained from these listings such as neighbourhood statistics, distribution of age around the area, and a list of closest schools with distance. Overall, 3 dataset was obtained from domain. Roughly 15,000 instances was obtained from scraping domain.

In [1]:
import pandas as pd
import pickle

In [3]:
# raw listing data from domain
listings = pd.read_csv('../data/raw/listing_properties.csv')
listings

Unnamed: 0  beds  baths  parking             title             price  \
0               0     1      1        0              $330              $330   
1               1     2      1        1           $430 pw           $430 pw   
2               2     4      3        2              $585              $585   
3               3     3      1        1              $420              $420   
4               4     1      1        1     $640 per week     $640 per week   
...           ...   ...    ...      ...               ...               ...   
13748       13748     4      2        2  $400 Per Week pw  $400 Per Week pw   
13749       13749     1      1        0     $435 per week     $435 per week   
13750       13750     3      2        4  $350 Per Week pw  $350 Per Week pw   
13751       13751     4      2        1              $600              $600   
13752       13752     2      1        0  $750/W Furnished  $750/W Furnished   

                                                 address promoType  \
0                     409/757 Bourke, Docklands VIC 3008     basic   
1      38/60-68 Gladesville Boulevard, Patterson Lake...   branded   
2              1 Liverpool Circuit, Craigieburn VIC 3064   branded   
3             8 Stoneman Turn, Caroline Springs VIC 3023   branded   
4            1510/1 Freshwater Place, Southbank VIC 3006  platinum   
...                                                  ...       ...   
13748                      7 Ebony Way, Tarneit VIC 3029     basic   
13749             17/416 St Kilda Rd, Melbourne VIC 3004     basic   
13750               16 Goulburn Court, Werribee VIC 3030     basic   
13751                 18 Jessie Street, Preston VIC 3072     basic   
13752       2509/228 La Trobe Street, Melbourne VIC 3000     basic   

      listingType             propertyType           status  mode  isRural  \
0            rent  Apartment / Unit / Flat  recentlyUpdated  rent    False   
1            rent  Apartment / Unit / Flat              new  rent    False   
2            rent                    House  recentlyUpdated  rent    False   
3            rent                    House              new  rent    False   
4            rent  Apartment / Unit / Flat              new  rent    False   
...           ...                      ...              ...   ...      ...   
13748        rent                    House             live  rent    False   
13749        rent  Apartment / Unit / Flat             live  rent    False   
13750        rent                    House             live  rent    False   
13751        rent                    House             live  rent    False   
13752        rent  Apartment / Unit / Flat             live  rent    False   

       houses    bond        id                  availableFrom  landArea  \
0           0  1433.0  14303265                            NaN       NaN   
1           0  1868.0  16089992         Monday, 28th June 2021       NaN   
2           0     NaN  15279569     Friday, 25th November 2022       NaN   
3           0  1820.0  16089977    Monday, 26th September 2022       NaN   
4           0  2781.0  16089972       Monday, 22nd August 2022       NaN   
...       ...     ...       ...                            ...       ...   
13748       0  1738.0  15434092   Wednesday, 27th October 2021       NaN   
13749       0  2500.0  14409092     Saturday, 9th October 2021       NaN   
13750       0  1521.0  15430756    Wednesday, 6th October 2021       NaN   
13751       0  2600.0  15428527      Tuesday, 5th October 2021       NaN   
13752       0  3258.0  15384018  Thursday, 15th September 2022       NaN   

       buildingArea method  
0               NaN    NaN  
1               NaN    NaN  
2               NaN    NaN  
3               NaN    NaN  
4               NaN    NaN  
...             ...    ...  
13748           NaN    NaN  
13749          50.0    NaN  
13750           NaN    NaN  
13751           NaN    NaN  
13752           NaN    NaN  

[13753 rows x 20 c

In [4]:
# suburb data from domain
sub_stat = pd.read_csv('../data/raw/suburb_stat.csv')
sub_stat

suburb  medianRentPrice  population    avgAge    owners  \
0             Docklands              380       10962  20 to 39  0.356243   
1       Patterson Lakes              430        7565  40 to 59  0.835075   
2           Craigieburn              450       50336  20 to 39  0.716903   
3      Caroline Springs              410       24200  20 to 39  0.766538   
4             Southbank              360       18706  20 to 39  0.347128   
...                 ...              ...         ...       ...       ...   
13711           Tarneit              400       34556  20 to 39  0.699615   
13712         Melbourne              375           0       NaN  0.000000   
13713          Werribee              340       40348  20 to 39  0.679556   
13714           Preston              600       32856  20 to 39  0.588015   
13715         Melbourne              450       47279  20 to 39  0.302987   

        renters  families   singles        id  
0      0.643757  0.393689  0.606311  14303265  
1      0.164925  0.527570  0.472430  16089992  
2      0.283097  0.556267  0.443733  15279569  
3      0.233462  0.571566  0.428434  16089977  
4      0.652872  0.312419  0.687581  16089972  
...         ...       ...       ...       ...  
13711  0.300385  0.611806  0.388194  15434092  
13712  0.000000  0.000000  0.000000  14409092  
13713  0.320444  0.463697  0.536303  15430756  
13714  0.411985  0.399137  0.600863  15428527  
13715  0.697013  0.237154  0.762846  15384018  

[13716 rows x 9 columns]

In [5]:
# neighbourhood data from domain
neigh_stat = pd.read_csv('../data/raw/neigh_stat.csv')
neigh_stat

age0To19  age20To39  age40To59  age60Plus  longTermResident     owner  \
0      0.129199   0.642119   0.195090   0.033592          0.136929  0.295322   
1      0.210918   0.248139   0.332506   0.208437          0.478632  0.639535   
2      0.328173   0.229102   0.346749   0.095975          0.000000  0.800000   
3      0.331512   0.251366   0.320583   0.096539          0.604072  0.777778   
4      0.093800   0.356121   0.356121   0.193959          0.202410  0.537366   
...         ...        ...        ...        ...               ...       ...   
13711  0.362153   0.427406   0.205546   0.004894          0.000000  0.374233   
13712  0.113907   0.540397   0.255629   0.090066          0.240545  0.355030   
13713  0.233898   0.277966   0.315254   0.172881          0.642857  0.663366   
13714  0.208772   0.356140   0.231579   0.203509          0.662698  0.641304   
13715  0.058824   0.786928   0.120261   0.033987          0.129870  0.154839   

         renter    family    single        id  
0      0.704678  0.512945  0.487055  14303265  
1      0.360465  0.576052  0.423948  16089992  
2      0.200000  0.647887  0.352113  15279569  
3      0.222222  0.600000  0.400000  16089977  
4      0.462633  0.568019  0.431981  16089972  
...         ...       ...       ...       ...  
13711  0.625767  0.602015  0.397985  15434092  
13712  0.644970  0.419152  0.580848  14409092  
13713  0.336634  0.570732  0.429268  15430756  
13714  0.358696  0.493088  0.506912  15428527  
13715  0.845161  0.171617  0.828383  15384018  

[13716 rows x 10 columns]

2. This data was further cleaned after it was obtained. There were many missing rows as those website listings has not been populated with data yet as of the time of scraping, and were dropped as a result. The rest of the cleaning process follows smoothly with simple tasks such as typecasting columns and splitting information such as the Suburb/Postcode from the address data. The dataset provided here allowed us to create visualisation and completed certain analysis on the different features at a listing level. For example, the correlation between the features provided was investigated to see if there was any interesting moving trend in relationships. Below is correlation for 2 bed flats to give a general idea on the correlations. Other properties displayed similar correlations

<img src="../plots/1_bed_flat_corr.png">

3. Suburb level data was obtained from the ABS. These include Median Rental Price per week from 2000 up until 2021, socio-economic-data for 2011,2016,2021 and Age distribution data. Median Rental Price was truncated to 2011-2021 as we planned to use 10 years worth of data for training our model, this dataset was also split into 6 different categories inorder to control for property types as a confounding variable on median price. In terms of Socio-Economic data, since the ABS only collected this information on a 5 year basis, we had alot of missing data for the years that are not 2011,2016,2021. Simple Linear Regression Imputation was chosen as the imputation technique over other simpler techniques such as median or mean imputation. This was chosen as there seems to be an increasing linear trend in socio-economic-features (and its a known fact that features such as Median personal Income increases overtime) and thus in theory Regression Imputation might be able to retain some of the true variation in data for the missing values, as compared to other techniques. Simple Regression was chosen over Multiple Regression as it was deemed unnecessary to impute a data base on more than just years as the regressor. Multiple Regressor would potentially introduce more noise into the data since randomise vector of initial inputs would be required to kickstart our imputation phase. Suburbs with no available data for any years were removed as linear imputation was not possible.

In [8]:
lin_imputed = pd.read_csv('../data/curated/all_imputed_monetary_data.csv')
lin_imputed

Median_age_persons_2011  Median_age_persons_2012  \
0                       26.0                25.866667   
1                       36.0                36.266667   
2                       30.0                29.933333   
3                       34.0                34.133333   
4                       29.0                29.200000   
..                       ...                      ...   
657                     40.0                41.533333   
658                     47.0                47.866667   
659                     39.0                39.733333   
660                     44.0                44.933333   
661                     49.0                49.666667   

     Median_age_persons_2013  Median_age_persons_2014  \
0                  26.066667                26.266667   
1                  36.866667                37.466667   
2                  30.033333                30.133333   
3                  34.433333                34.733333   
4                  29.400000                29.600000   
..                       ...                      ...   
657                41.733333                41.933333   
658                48.066667                48.266667   
659                40.133333                40.533333   
660                45.533333                46.133333   
661                50.166667                50.666667   

     Median_age_persons_2015  Median_age_persons_2016  \
0                  26.466667                     26.0   
1                  38.066667                     38.0   
2                  30.233333                     30.0   
3                  35.033333                     35.0   
4                  29.800000                     30.0   
..                       ...                      ...   
657                42.133333                     45.0   
658                48.466667                     50.0   
659                40.933333                     42.0   
660                46.733333                     48.0   
661                51.166667                     52.0   

     Median_age_persons_2017  Median_age_persons_2018  \
0                  26.866667                27.066667   
1                  39.266667                39.866667   
2                  30.433333                30.533333   
3                  35.633333                35.933333   
4                  30.200000                30.400000   
..                       ...                      ...   
657                42.533333                42.733333   
658                48.866667                49.066667   
659                41.733333                42.133333   
660                47.933333                48.533333   
661                52.166667                52.666667   

     Median_age_persons_2019  Median_age_persons_2020  ...  \
0                  27.266667                27.466667  ...   
1                  40.466667                41.066667  ...   
2                  30.633333                30.733333  ...   
3                  36.233333                36.533333  ...   
4                  30.600000                30.800000  ...   
..                       ...                      ...  ...   
657                42.933333                43.133333  ...   
658                49.266667                49.466667  ...   
659                42.533333                42.933333  ...   
660                49.133333                49.733333  ...   
661                53.166667                53.666667  ...   

     Average_household_size_2014  Average_household_size_2015  \
0                       1.853333                     1.843333   
1                       1.833333                     1.833333   
2                       2.160000                     2.130000   
3                       1.840000                     1.820000   
4                       1.886667                     1.876667   
..                           ...                          ...   
657                     2.653333                     2.643333   


4. More external datasets were then obtained along the way, namely Crime Statistics and Population data for the prediction model. As Crime Statistics was available between 2013-2021, we have decided to further reduce our training timeframe from 2011-2021 to 2013-2021. The two datasets were then joined with the aforementioned time series data. After removing all categorical variables, Pearson's Correlation matrix was created to visualise any strong moving trends between features and Median Rental Price. This was used with our own judgement to further remove unnecessary features from our prediction dataset. The final prediction dataset contains Median_age_persons, Median_tot_prsnl_inc_weekly, Average_household_size, Population, Alleged Offender Incident Rate, Victimization Rate, Offender Count.

In [12]:
crime = pd.read_csv('../data/curated/Crime_Rate_Suburb/crime_rate_all_yrs.csv')
crime

Suburb           LGA Alleged_Offender_Incident_Rate_2013  \
0        Abbeyard        ALPINE                            1,359.60   
1       Barwidgee        ALPINE                            1,359.60   
2          Bogong        ALPINE                            1,359.60   
3          Bright        ALPINE                            1,359.60   
4        Buckland        ALPINE                            1,359.60   
...           ...           ...                                 ...   
2880      Preston       DAREBIN                            2,211.60   
2881    Reservoir       DAREBIN                            2,211.60   
2882    Thornbury       DAREBIN                            2,211.60   
2883  Queenscliff  QUEENSCLIFFE                               899.4   
2884  Swan Island  QUEENSCLIFFE                               899.4   

     Alleged_Offender_Incident_Rate_2014 Alleged_Offender_Incident_Rate_2015  \
0                               1,618.80                            1,512.60   
1                               1,618.80                            1,512.60   
2                               1,618.80                            1,512.60   
3                               1,618.80                            1,512.60   
4                               1,618.80                            1,512.60   
...                                  ...                                 ...   
2880                            2,175.80                            2,161.60   
2881                            2,175.80                            2,161.60   
2882                            2,175.80                            2,161.60   
2883                               976.4                            1,261.50   
2884                               976.4                            1,261.50   

     Alleged_Offender_Incident_Rate_2016 Alleged_Offender_Incident_Rate_2017  \
0                               1,470.80                            1,186.90   
1                               1,470.80                            1,186.90   
2                               1,470.80                            1,186.90   
3                               1,470.80                            1,186.90   
4                               1,470.80                            1,186.90   
...                                  ...                                 ...   
2880                            2,382.60                            2,553.70   
2881                            2,382.60                            2,553.70   
2882                            2,382.60                            2,553.70   
2883                            1,433.90                            1,419.40   
2884                            1,433.90                            1,419.40   

     Alleged_Offender_Incident_Rate_2018 Alleged_Offender_Incident_Rate_2019  \
0                               1,547.50                            1,693.70   
1                               1,547.50                            1,693.70   
2                               1,547.50                            1,693.70   
3                               1,547.50                            1,693.70   
4                               1,547.50                            1,693.70   
...                                  ...                                 ...   
2880                            2,264.10                            2,203.10   
2881                            2,264.10                            2,203.10   
2882                            2,264.10                            2,203.10   
2883                               704.2                            1,463.10   
2884                               704.2                            1,463.10   

     Alleged_Offender_Incident_Rate_2020  ... Victimisation_Rate_2021  \
0                               1,896.50  ...                1,469.10   
1                               1,896.50  ...                1,469.10   
2                               1,896.50  ...           

In [14]:
pop = pd.read_csv('../data/curated/Pop_imputed_2011-2031.csv')
pop

2011   2012   2013   2014   2015   2016   2017   2018   2019   2020  \
0    21566  22987  25655  28324  30992  33660  36329  38997  41665  44334   
1      871    755   3246   5737   8227  10718  13209  15699  18190  20681   
2     1419  -1730    463   2657   4852   7046   9240  11435  13629  15823   
3     9054  10506  12404  14301  16198  18096  19993  21891  23788  25685   
4     1894   2006   3886   5766   7646   9526  11406  13286  15166  17046   
..     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
477  14955  14836  14817  14799  14781  14762  14744  14725  14707  14689   
478   8575   8540   8518   8496   8474   8452   8430   8409   8387   8365   
479   8939   8870   8846   8821   8797   8772   8748   8723   8699   8674   
480   6957   6891   6863   6835   6807   6779   6751   6723   6695   6667   
481  38357  38187  38083  37979  37874  37770  37666  37562  37458  37354   

     ...   2023   2024   2025   2026   2027   2028   2029   2030   2031  \
0    ...  52338  55007  57675  60343  63012  65680  68348  71017  74932   
1    ...  28153  30644  33134  35625  38116  40606  43097  45588  50685   
2    ...  22406  24600  26795  28989  31183  33378  35572  37766  45305   
3    ...  31378  33275  35172  37070  38967  40865  42762  44659  47002   
4    ...  22686  24566  26446  28326  30206  32087  33967  35847  39495   
..   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
477  ...  14633  14615  14597  14578  14560  14541  14523  14505  14587   
478  ...   8299   8277   8255   8233   8211   8190   8168   8146   8137   
479  ...   8601   8576   8552   8527   8503   8478   8454   8429   8449   
480  ...   6584   6556   6528   6500   6472   6444   6416   6388   6399   
481  ...  37042  36938  36834  36730  36626  36522  36418  36314  36276   

                Suburb  
0              Tarneit  
1              Wollert  
2             Rockbank  
3         Melton South  
4              Officer  
..                 ...  
477     Meadow Heights  
478         Kings Park  
479     Gladstone Park  
480  Aspendale Gardens  
481   Hoppers Crossing  

[482 rows x 22 columns]

5. External datasets Train Stations, Shopping Centers, Universities, Parks, Primary/Secondary Schools were also obtained with the help of OpenStreetMaps API, external open data-sources namely DataVIC or straight from domain. The proximity to these features at instance level was calculated using OpenRouteServices. Any proximity that was >= 100km to the nearest places of interest was omitted as it was assumed that the residents would not be utilising these facilities to the point that it becomes a factor in affecting the rental price. The name of the nearest schools was also obtained since there might be higher demand in the area where the school is known for high reputations. Below is the correlation for 2 bed flats

<img src="../plots/bed_2_flat.png"/>

To breifly summarise the correlations, proximity to nearest places of interest generally shows a negative relationship with the rental price, (the price goes down as the proximity increases). 
One interesting relationship shown here is that the distribution of age 20-39. The higher the ratio, the higher the rental prices



6. The features were also fit into a  Linear Regresison model along with the neighbourhood data of each listings. The closest shopping centres, stations and universities were omitted as they showed a strong correlation with the Suburb, which would result in multi-collinarity. Backwards elimiantion was used to filter out the features that did not show much significance in predicting the rental price. Below shows one example of the result of anovas, the features listed corresponds to one of the most important features in determining the rental price at an instance level

In [11]:
pd.read_csv("../data/curated/bed_2_flat_anova.csv")

Unnamed: 0        sum_sq      df          F        PR(>F)
0          Suburb  2.203627e+06   189.0   2.125201  6.579525e-15
1  closest_school  4.741030e+06   558.0   1.548682  2.642380e-11
2        age0To19  7.436388e+04     1.0  13.554573  2.390201e-04
3       age20To39  6.508158e+04     1.0  11.862655  5.867218e-04
4           owner  2.716764e+04     1.0   4.951945  2.619311e-02
5        Residual  9.255316e+06  1687.0        NaN           NaN

As a general observation we see that rather than the proximity, the closest school itself had more importance in determining the rental price of the listing. The age demographic of the area also seems to show some strong form of relationship however this may be correlation rather than causation

## Prediction of rental prices 2020-2030

1. Support Vector Regression was the first that came to our mind as the model of choice since our data has many dimensions, and doesn't exactly have a huge number of instances (roughly 90 suburbs after cleaning). SVR can also predict continuous data and thus was perfect for Median Rental Price prediction. We attempted to hyperparameter tune our model, however, as GridSearchCV was extremely computationally expensive, we arrived at a hyperparameter set that wasnt obtained through exhaustive search.

In [17]:
with open('../data/curated/forecast_bed_1_flat_covid', 'rb') as f:
    forecast_bed_1_flat_covid = pickle.load(f)
with open('../data/curated/forecast_bed_2_flat_covid', 'rb') as f:
    forecast_bed_2_flat_covid = pickle.load(f)
with open('../data/curated/forecast_bed_2_house_covid', 'rb') as f:
    forecast_bed_2_house_covid = pickle.load(f)

In [19]:
forecast_bed_2_house_covid['Collingwood, Collingwood North']

Median_age_persons  Median_mortgage_repay_monthly  \
0                 32.8                         2170.5   
1                 32.9                         2170.1   
2                 33.0                         2169.6   
3                 33.1                         2169.2   
4                 33.2                         2168.8   
5                 33.3                         2168.3   
6                 33.5                         2167.9   
7                 33.6                         2167.5   
8                 33.7                         2167.0   
9                 33.8                         2166.6   
10                33.9                         2166.2   

    Median_tot_prsnl_inc_weekly  Median_rent_weekly  \
0                        1259.2               423.4   
1                        1321.1               436.5   
2                        1382.9               449.6   
3                        1444.8               462.7   
4                        1506.7               475.7   
5                        1568.6               488.8   
6                        1630.5               501.9   
7                        1692.4               515.0   
8                        1754.2               528.1   
9                        1816.1               541.2   
10                       1878.0               554.2   

    Median_tot_fam_inc_weekly  Average_num_psns_per_bedroom  \
0                      2688.7                           1.0   
1                      2800.4                           1.0   
2                      2912.1                           1.0   
3                      3023.9                           1.0   
4                      3135.6                           0.9   
5                      3247.3                           0.9   
6                      3359.1                           0.9   
7                      3470.8                           0.9   
8                      3582.5                           0.9   
9                      3694.3                           0.8   
10                     3806.0                           0.8   

    Average_household_size      Pop  Alleged_Offender_Incident_Rate  \
0                      1.9   8464.0                          3747.7   
1                      1.9   8637.3                          3760.2   
2                      1.9   8810.5                          3772.6   
3                      1.9   8983.7                          3785.0   
4                      1.9   9157.0                          3797.4   
5                      1.9   9330.2                          3809.8   
6                      1.9   9503.4                          3822.2   
7                      1.8   9676.7                          3834.6   
8                      1.8   9849.9                          3847.0   
9                      1.8  10023.1                          3859.4   
10                     1.8  10196.4                          3871.8   

    Victimisation_Rate  Offence_Count  year  Median_rental_price  
0               6583.3        14030.4  2020           578.866202  
1               6557.6        14375.1  2021           586.840529  
2               6531.9        14719.9  2022           598.518693  
3               6506.2        15064.6  2023           614.447998  
4               6480.4        15409.4  2024           635.376556  
5               6454.7        15754.1  2025           662.140181  
6               6429.0        16098.9  2026           702.527990  
7               6403.3        16443.6  2027           851.758125  
8               6377.6        16788.4  2028           919.079375  
9               6351.8        17133.1  2029           995.882108  
10              6326.1        17477.9  2030          1083.266989

6. SVR gave us Median Rental Price from 2022-2027 as specified, this forecasted data, including the forecasted data for the features were used to create visualisation and analysis shown in the presentation slide.

### Visualising the predictions
In order to better visualise those predictions, a way of putting those prediction into a map was created. The function allows the user to choose the type of property to show on the cloropleth, and the year. Predictions for 2027 for the 1 bedroom flat, and 2 bedroom houses and flats have been generated and placed in the `plots` folder.

From the visualisation we can see our problems with missing data, many suburbs have been left out from the predictions, as there was too much data missing.

![2 bed flat](../plots/2_bed_flat.png)

## Identifying/ranking liveability

1. The liveability data was defined using the following metrix with each weighted as follows:
* Crime rate of the suburb - 0.3
* Proximity to the nearest train station  - 0.20
* Number of schools in the suburb - 0.20
* Number of parks, shopping centres in the suburb - 0.15, 0.15

Each suburb was given a score based on this weight and ordered, which produced the ranking shown in the presentation

Note: The metrics were first standardised and the reasoning for choice of each weight was explained in the presentation

